Informática - 1º de Física
<br>
**Introducción a la Programación**
<br>
<p style="color:#808080"> <small><small>
18/09/2018
</small></small></p>

## Tema 1. Primeros pasos

El punto de partida es instalar en nuestro ordenador el lenguaje de programación que vamos a utilizar y algunas herramientas auxiliares. Una vez hecho esto, existen diferentes formas de trabajar:

- Editor y terminal.

- IDE (entorno integrado de programación).

- Notebook interactivo.

En su momento haremos ejercicios para familiarizarnos con las dos primeras pero la mayoría de las explicaciones las haremos utilizando los notebooks. De esta manera podemos crear documentos de referencia con todo lo que vamos aprendiendo.

A partir de aquí suponemos que conocemos los comandos de edición básicos de los notebooks, que se explican en la primera clase de prácticas. (También se puede echar un vistazo a la opción "Help -> Notebook Help" de la barra de menú superior.)

Se recomienda crear una carpeta en tu ordenador para guardar todos los documentos que se vayan creando en la asignatura.

### aritmética elemental

Igual que una calculadora, un ordenador es capaz de hacer operaciones aritméticas:

In [1]:
3+2

5

Los símbolos de las operaciones son más o menos intuitivos: `+`, `-`, `*`, `/`, y la precedencia es la usual en matemáticas.

In [2]:
2 + 3 * 5

17

Los espacios no influyen. Usamos paréntesis para agrupar de la forma deseada:

In [3]:
(2+3) *  5-4

21

La potencia se expresa con el símbolo `**`.

In [4]:
2**10

1024

En programación es muy importante distinguir entre números enteros y números reales. Los enteros son matemáticamente exactos, mientras que los reales son aproximaciones con una cantidad finita de decimales.

In [5]:
2**(1/2)

1.4142135623730951

En algunos lenguajes los enteros tienen un tamaño limitado, como ocurre en las calculadoras. Pero en Python los enteros pueden ser de cualquier tamaño. El único límite es la memoria disponible en el ordenador.

In [6]:
13**200

6147102592468651336192412089282069174074558674790952240521262267958019995940288912789820651282318338938794713404346722691907328564818184457431599939243408750167870757463938171866175799570066146915175087199725900885179188001

El símbolo de división `/` se refiere a la división real:

In [7]:
10 / 2

5.0

In [8]:
11 / 2

5.5

El cociente y el resto de la división entera utilizan respectívamente los símbolos `//` y `%`: 

In [9]:
11 // 2

5

In [10]:
11 % 2

1

Los números reales que se utilizan normalmente son aproximaciones de "coma flotante" de "doble precisión" (64bits) que tienen 15 decimales y un exponente hasta $\pm 308$.

La notación científica $a\times 10^b$ se expresa $a$ `e` $b$. 

In [11]:
2.37e18 ** 5.1

5.143000477600881e+93

Debido a que los números reales tienen un número fijo de decimales las operaciones pueden producir errores de redondeo:

In [12]:
10/3

3.3333333333333335

In [13]:
0.1 + 0.2

0.30000000000000004

### bibliotecas de funciones

El lenguaje Python posee un número enorme de funciones y operaciones disponibles que están organizadas en una gran biblioteca estructurada por "módulos". Por ejemplo, las funciones matemáticas (trigonométricas, logaritmos, etc.) se guardan en el módulo `math`.

Para utilizar funciones de un módulo debemos "importarlas" (hacerlas visibles) con una instrucción como la siguiente:

In [14]:
from math import pi, sin, sqrt

In [15]:
sin(pi/6)

0.49999999999999994

Se pueden importar todas de una vez haciendo `from math import *` pero no se recomienda. La forma adecuada de acceder a todas las funciones de un módulo es utilizar un prefijo:

In [16]:
import math

In [17]:
math.cos(0)

1.0

In [18]:
import math as m

In [19]:
m.log(10)

2.302585092994046

Es muy conveniente estar familiarizado con las funciones disponibles en los módulos relacionados con nuestro ámbito de aplicación (principalmente `numpy`, `scipy`, `matplotlib` y `pandas`). No hay que aprender nada de memoria sino tener a mano la documentación, que está disponible online  (p. ej. [aquí tenemos](
https://docs.python.org/3/library/math.html) las funciones disponibles en `math`) y localmente con la función `help`. En el menú "Help" del notebook hay enlaces a los módulos más importantes. En cualquier caso una búsqueda en google nos lleva rápidamente a la función deseada.

Con lo visto hasta ahora podemos usar el lenguaje Python como una calculadora. A partir de aquí iremos introduciendo la capacidad de programar las operaciones. 

### secuencia de instrucciones

Python evalúa una detrás de otra todas las expresiones que encuentra. Pero el notebook solo muestra el resultado de la última que aparece en la celda:

In [20]:
1 + 1

2 + 3

5 * 2

10

Si queremos ver todos los resultados utilizamos la función `print`:

In [21]:
print(1 + 1)

print(2 + 3)

print(5 * 2)

2
5
10


### nombres

Un aspecto fundamental en la programación consiste en ponerle nombres a los números y otros tipos de datos con los que estamos trabajando. Para ello, Python (y la mayoría de lenguajes) usa el símbolo `=`. Por ejemplo, si hacemos lo siguiente

In [22]:
a = 5

z = 1/2 + 3*a

entonces los nombres `a` y `z` queda definidos con los siguientes valores:

In [23]:
a

5

In [24]:
z

15.5

La elección del signo `=` no es muy afortunada, ya que esta acción no tiene mucho que ver con el concepto de igualdad matemática. (Otros lenguajes usan signos más intuitivos como `<-` o `:=` para esta operación.)

Cuando veamos expresiones como `z = 1/2 + 3*a`  no tenemos que pensar en ningún momento en ecuaciones matemáticas. Nada más lejos de la realidad. Simplemente se da un nombre al **resultado** de evaluar la expresión. Podríamos decir que estamos "definiendo" un nombre, pero este término también puede inducir a confusión, ya que el nombre queda asignado al valor final de la expresión, y no a la expresión en sí. Cada vez que aparezca `z` en el programa se sustituirá automáticamente por `15.5`. No se sustituye por `1/2 + 3*a`. 

Este matiz es importante debido a que los nombres se pueden redefinir. Observa la siguiente secuencia de definiciones:

In [25]:
a = 5

b = 2*a

a = 7

b

10

Recuerda que las expresiones se evalúan una detrás de otra. La última de ellas, `b` tiene el valor `10` a pesar de que se ha definido como `b = 2*a` y que justo antes se ha redefinido `a` como `a=7`. El cambio en un nombre no afecta a los demás que dependen de él. No se recalcula nada. 

Más adelante explicaremos cómo hacer verdaderas definiciones en las que unos símbolos dependen de otros. Por el momento podemos pensar en la idea de almacenar en memoria un dato que podemos recuperar cómodamente mediante un nombre.

En algunos lenguajes se dice que hemos "almacenado un valor en una variable", o que hemos "asignado un valor". Tampoco es lo ideal porque la palabra "variable" puede inducir a confusión con las variables matemáticas. (Y no causa buena impresión almacenar constantes como por ejemplo el número $\pi$ dentro variables...)

No hay ningún problema en redefinir una variable usando su valor anterior. Esto puede dar lugar a expresiones aparentemente absurdas como:

In [26]:
b = b + 5

Simplemente tenemos que recordar que lo anterior no es una ecuación sino una redefinición, cuyo efecto es sustituir del valor de `b`, que antes era `10`, por `15`. 

Esta operación se puede abreviar como `b += 5`.

Cuando se redefine un nombre el valor que tenía previamente se olvida de forma irrecuperable.

Para hacer una definición el nombre (cualquier palabra que empiece por una letra) se pone a la izquierda del signo `=` y a la derecha se pone la expresión deseada, donde pueden aparecer nombres previamente definidos. Por tanto, lo siguiente es incorrecto:

In [27]:
1 + 1 = c

SyntaxError: can't assign to operator (<ipython-input-27-a6067e38b1ca>, line 1)

In [49]:
c = 3*x + 5

NameError: name 'x' is not defined

Es conveniente elegir nombres descriptivos o abreviaturas para indicar el papel que juega cada dato en el programa.

In [50]:
masa = 10
velocidad_inicial = 8.5
pos = 5

α  = pi/4

Es una cuestión de gusto personal, pero siempre se recomienda la claridad y la sencillez.

### estado de la computación

Cuando el ordenador "ejecuta" un programa (o Python evalúa celdas de código) ocurre lo siguiente:

- En todo momento existe un *estado* de la computación, que es simplemente un conjunto de datos con sus correspondientes nombres (inicialmente vacío).

- Algunas instrucciones cambian elementos del estado. Otras no, aunque sí actúan sobre el "mundo exterior" (por ejemplo, mostrando algo en la pantalla). El efecto producido por una instrucción depende del estado anterior.

Una computación es la secuencia de estados producida por las sucesivas instrucciones del programa. El *resultado* de la computación es el estado final.

Por ejemplo:

In [2]:
# Instrucción       Estado
                #   {}
a = 5
                #   { a:5 }
b = 2*a
                #   { a:5 , b:10 }
print(b)
                #   { a:5 , b:10 }
a = 7
                #   { a:7 , b:10 }
b += 2
                #   { a:7 , b:12 }
a+b
                #   { a:7 , b:12 }
a-b
                #   { a:7 , b:12 }

10


-5

La instrucción `print` de la línea 7 escribe `10` en la pantalla (o notebook). La línea 13 realiza una operación pero no tiene ningún efecto observable. El resultado se descarta. La línea 15 sí imprime el resultado `-5` por ser la última (esto es un convenio de los notebooks para ahorrar la instrucción `print`).

### tipos de datos

Un concepto esencial en programación es el de **tipo** de un dato. Intuitivamente, podemos pensar en un conjunto de valores. Un buen ejemplo son los tipos numéricos, que tratan de imitar a los conjuntos matemáticos usuales $\mathbb Z$, $\mathbb Q$, $\mathbb R$, $\mathbb C$, etc.

Hasta el momento hemos visto datos enteros (tipo `int`) y reales (tipo `float`) pero los ordenadores pueden 
trabajar con muchos otros tipos de información. La función `type` sirve para consultar de qué tipo es un dato:

In [51]:
type(b)

int

In [52]:
type(-2/3)

float

### números complejos

La operaciones con complejos están disponibles en el módulo [cmath](https://docs.python.org/3.6/library/cmath.html?highlight=cmath#module-cmath). Las constantes imaginarias se introducen añadiendo la letra `j`:

In [53]:
import cmath

In [54]:
3j ** 2

(-9+0j)

In [55]:
cmath.sqrt(-25)

5j

In [56]:
a = 2+3j

cmath.cos(a)

(-4.189625690968807-9.109227893755337j)

In [57]:
cmath.cos(a)**2 + cmath.sin(a)**2

(1-1.4210854715202004e-14j)

In [58]:
type(a)

complex

### operaciones lógicas

 `>`, `<`, `>=`, `<=`, `==`, `!=`, `and`, `or`, `not`.

Un tipo de datos esencial en programación es el tipo `bool` (lógico), que solo puede tomar dos valores: `True` o `False`. Aparece como resultado de operaciones de comparación. Por ejemplo:

In [59]:
type(1 > 2)

bool

In [60]:
3 < 10

True

In [61]:
b - 3 == 8 + 7

False

Las operaciones de comparación se expresan con símbolos más o menos autoexplicativos: `>`, `<`, `>=`, `<=`, `==`, `!=` y, como veremos, se usan para tomar decisiones en los programas. Dados dos números comprueban si la condición es cierta o no.

Observa que el símbolo `==` que se usa para comprobar si dos números son iguales se escribe con dos signos de igual seguidos para distinguirlo del operador `=` de asignación.

A pesar de su aspecto, las operaciones de comparación no establecen una relación matemática entre las variables. La sentencia anterior no define una ecuación que el ordenador tenga que resolver. Es simplemente una expresión que se evalúa y produce un resultado que en este caso no es numérico sino lógico (verdadero o falso).

Los valores lógicos se pueden combinar mediante las conectivas `and`, `or`, `not` de la lógica proposicional. Por ejemplo:

In [62]:
3 < 4 or 2+2 == 5

True

In [63]:
b == 13 and (b < 0 or not 2+2==5)

False

La precedencia de los operadores lógicos es la usual: primero se evalúan las comparaciones, a continuación `not`, después `and` y finalmente `or`.

In [64]:
not True and False or True

True

In [65]:
not (True and (False or True))

False

Finalmente, es importante señalar que nunca se deben hacer comparaciones de igualdad exacta con números de tipo `float` aunque lo permita el lenguaje:

In [66]:
0.2 + 0.5 == 0.7

True

In [67]:
0.2 + 0.1 == 0.3

False

Este último resultado inesperado (y erróneo) se debe al carácter aproximado de los números reales informáticos y a la imposibilidad de representar exactamente muchos números racionales en la base de numeración binaria utilizada internamente.
Como veremos después, la comparación de números reales debe hacerse estableciendo una tolerancia sobre su diferencia absoluta o relativa, según convenga.

El procesador opera con los números de coma flotante de una forma muy razonable, con un compromiso entre precisión y velocidad adecuado para la mayoría de las aplicaciones. Pero olvidar la naturaleza aproximada de la computación numérica puede conducir a errores en el programa que se manifiestan cuando menos te lo esperas.




### cadenas de caracteres

El otro tipo de datos muy importante es la cadena de caracteres (`str`). Se trata simplemente de una secuencia de letras, números o cualquier otro símbolo. En las aplicaciones científicas se usa normalmente para mostrar mensajes al usuario humano, o para introducir la información a los programas. 

In [68]:
nombre = "Alberto"

saludo = 'Hola ' + nombre

saludo

'Hola Alberto'

In [69]:
type(nombre)

str

In [70]:
'El número pi es π = {:.4f}'.format(pi)

'El número pi es π = 3.1416'

Una cadena de caracteres es un simple texto entre comillas que en principio no tiene ningún significado para el ordenador. Más adelante veremos algunas operaciones útiles para manipular cadenas pero en este curso nos centraremos principalmente en operaciones numéricas.

Por comodidad algunas operaciones que en realidad son distintas se representan con el mismo símbolo. Por ejemplo, el signo `+` se utiliza para sumar diferentes tipo de números (lo cual implica un trabajo interno especial para cada caso) y también para concatenar cadenas.

Finalmente, vamos a presentar brevemente algunos tipos numéricos especializados para hacernos una idea de las herramientas disponibles para la computación científica.

Los apartados marcados con asterisco pueden omitirse en una primera lectura.

### racionales *

El módulo [fractions](https://docs.python.org/3.6/library/fractions.html) proporciona números racionales exactos. Son útiles en teoría de números pero no se utilizan mucho en ciencias experimentales.

In [2]:
from fractions import Fraction

In [3]:
x = Fraction(13,1001)

In [4]:
x

Fraction(1, 77)

In [5]:
Fraction(7,8)**17 + x

Fraction(20164349390700187, 173388585653764096)

### precisión arbitraria *

La aritmética de coma flotante con precisión arbitraria se puede conseguir (entre otros) con el módulo [mpmath](http://mpmath.org/). Esto permite elegir el número de decimales deseado en los cálculos.

In [45]:
from mpmath import mp
mp.dps = 100

(1+mp.sqrt(5))/2

mpf('1.618033988749894848204586834365638117720309179805762862135448622705260462818902449707207204189391137492')

La función `mp.mpf` construye números de este tipo para usarlos en la artimética normal. Pero hay que tener cuidado. Observa la diferencia:

In [47]:
mp.mpf(1/3)

mpf('0.333333333333333314829616256247390992939472198486328125')

In [48]:
mp.mpf(1)/3

mpf('0.3333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333333345')

En ciencias experimentales no necesitamos normalmente una precisión tan enorme. Además, estos números no llevan la cuenta del número de dígitos exactos que van quedando después de cada operación. 

### magnitudes físicas *

El módulo [uncertainties](http://pythonhosted.org/uncertainties/) automatiza la propagación de errores. Las magnitudes físicas se pueden manejar mediante [pint](http://pint.readthedocs.io/en/0.8/tutorial.html). Estas herramientas son muy útiles para analizar datos de los laboratorios de física. Más adelante veremos algunos ejemplos de uso.

### Ejercicios

- La [razón áurea](https://en.wikipedia.org/wiki/Golden_ratio) es el número

 $$g = \frac{1+\sqrt{5}}{2} $$

 Evalúa $g$ usando los números de coma flotante estándar de doble precisión (`float`) y comprueba si la aproximación obtenida verifica la ecuación 

 $$\frac{g+1}{g} = g$$


- Escribe expresiones para transformar grados [Fahrenheit](https://en.wikipedia.org/wiki/Fahrenheit) en centígrados y comprueba su funcionamiento con unos cuantos valores.

 $$C = \frac{F-32}{1.8}$$

 (No te preocupes por ahora por la entrada de datos. Puedes almacenar directamente el valor de interés en una variable, o si quieres puedes usar la función `input`.)


- Escribe expresiones para calcular las soluciones de una ecuación de segundo grado y comprueba su funcionamiento en unos cuantos casos de prueba.


- Expresa en Python la fórmula del período de un péndulo de longitud $l$ (para el caso de oscilaciones pequeñas) y evalúala para $l=1m$ y $l=5cm$ .

 $$ T_0 = 2\pi \sqrt{\frac{l}{g}} $$

 En general, si la amplitud de las oscilaciones es $\phi$, el período puede aproximarse mediante: 
 
 $$ T = T_0\left(1 + \frac{1}{16}\phi^2 + \frac{11}{3072}\phi^4  + \ldots \right)$$

 Compara el tamaño de la corrección para oscilaciones de $\phi=10$ y $\phi=30$ grados.



- Utiliza las operaciones numéricas de tipo `complex` para comprobar que

 $$e^{i\pi}+1=0$$


- Escribe un programa para resolver la [ecuación de tercer grado](https://en.wikipedia.org/wiki/Cubic_function).